### PR TITLE
Supports snakecase.

### LINKED ISSUE #170: Please support snake case
See https://gist.github.com/agemooij/7679130

### PR DESCRIPTION
**Pull request by jkugiya**
Close #170

### CODE REVIEW

**Comment by morpheyesh** [comment]
Hope this will be merged soon!

File: src/main/scala/spray/json/SnakeCaseJsonSupport.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import scala.reflect.ClassTag
+import scala.collection.mutable
+
+/**
+  * Provides snake cased JsonFormats
+  */
+trait SnakeCaseJsonSupport extends DefaultJsonProtocol {
+
+  override protected def extractFieldNames(classTag: ClassTag[_]): Array[String] = {
+    def snakify(name: String) = {
```
Comment (jrudolph) [inline src/main/scala/spray/json/SnakeCaseJsonSupport.scala]: Maybe we should also just use the regex here? Performance shouldn't be an issue as the formats are usually only evaluated once per type.

File: src/test/scala/spray/json/SnakeCaseJsonSupportSpec.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      implicit val jsonFormat1: RootJsonFormat[SampleFields] =
+         jsonFormat21(SampleFields.apply)
+      implicit val jsonFormat2: RootJsonFormat[SampleFields2] =
+        jsonFormat7(SampleFields2.apply)
+      // When
+      val sampleFields1 = SampleFields().toJson.asJsObject
+      val sampleFields2 = SampleFields2().toJson.asJsObject
+      // Then
+      val fields1 = sampleFields1.fields
+      fields1.contains(snakify("HelloWorld")) mustEqual true
```
Comment (jrudolph) [inline src/test/scala/spray/json/SnakeCaseJsonSupportSpec.scala]: Good that `snakify` is a different implementation here in the test, but I would prefer if the test case expectations would be spelled out.

Comment (jrudolph) [inline src/test/scala/spray/json/SnakeCaseJsonSupportSpec.scala]: Good that `snakify` is a different implementation here in the test, but I would prefer if the test case expectations would be spelled out.

Comment (jrudolph) [reply src/test/scala/spray/json/SnakeCaseJsonSupportSpec.scala]: E.g. like this:

```suggestion
      fields1 must haveKey("hello_world")
```

**Comment by plokhotnyuk** [comment]
@jrudolph Regexs are so slow... 

More simple and efficient functions which can enforce kebab case too are [here](https://github.com/plokhotnyuk/jsoniter-scala/blob/3667262680d6ee527058dbd8cbebdacdec305c97/jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala#L114-L151)

**Comment by jkugiya** [comment]
> @jrudolph Regexs are so slow...

Yes, the reason for this implementation is regex's performance. But I can understand regex is easier to maintain. I created another PR(#295). I'll leave it up for either.

**Comment by jkugiya** [comment]
@plokhotnyuk I also tried your implementation. But some of the behavior is [unexpected for me](https://github.com/jkugiya/spray-json/blob/feat/support_snakecase_simple_impl/src/test/scala/spray/json/SnakeCaseJsonSupportSpec.scala#L50-L78) in some corner case.

ex)
- `^There-arE_Non_(Alphabetic))_character!S` -> `^_there_-ar_e__non_(_alphabetic_))_character_!s` (The capital letter is followed by `_`.)
- -> `  There are space character ` -> `  _there_ are_ space_ character_ ` (it start with `_`)

It is difficult to judge which is better because there aren't standard specification for this conversion, as far as I know.
